driver-sync/src/main/com/mongodb/client/CommonListCollectionsIterable.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client;
+
+import com.mongodb.lang.Nullable;
+import org.bson.BsonValue;
+import org.bson.conversions.Bson;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This interface is shared by {@link ListCollectionsIterable}, {@link ListCollectionNamesIterable}.
+ * Each of those interfaces may have additional methods specific to it.
+ */
+interface CommonListCollectionsIterable<TResult> extends MongoIterable<TResult> {
```

Comment:
I think it might be surprising for users that this non-public interface is the path through which you get to the public `MongoIterable` interface. It might be clearer if this interface didn't extend `MongoIterable` and instead `ListCollectionNamesIterable` and `ListCollectionsIterable` extend both this and `MongoIterable`. 

I'd also be ok with removing this interface entirely, as it introduces a new pattern that is not followed for the other iterables.  For example, `FindIterable` and `AggregateIterable` also declare all the methods in this interface, but do not extend it.

Reply:
> I think it might be surprising for users that this non-public interface is the path through which you get to the public MongoIterable interface.

The interface is package-access, so it's not visible in the API documentation. The screenshot from the docs I generated locally:

<img width="548" alt="Screenshot 2023-12-06 at 09 14 50" src="https://github.com/mongodb/mongo-java-driver/assets/6304193/bac85602-07b6-48bd-8af9-e6723f00121d">

This interface is simply an internal detail that makes sure that the methods that must be present on both `ListCollectionsIterable`, `ListCollectionNamesIterable` are indeed present there.

> `FindIterable` and `AggregateIterable` also declare all the methods in this interface, but do not extend it.

Maybe they should (that will require renaming the interface into something like `ExtendedMongoIterable`)? Or maybe these methods should go into `MongoIterable`, if they are shared by all iterables we have (I suspect this is not the case).

Reply:
> The interface is package-access, so it's not visible in the API documentation.

I see.  But I think that most people explore the API via Command-B in IDEA or equivalent, rather than finding the Javadoc on the web.  Does that change your perspective at all?

Reply:
> I think that most people explore the API via Command-B in IDEA or equivalent, rather than finding the Javadoc on the web. Does that change your perspective at all?

Not in the context of this discussion.

However, I vaguely remember that Slava encountered a similar problem in how our iterable interfaces are organized: they duplicate methods with no mechanism enforcing the duplicated sets of methods to stay consistent. I contacted him, and turned out, he even created a corresponding ticket https://jira.mongodb.org/browse/JAVA-5068. So what I'll do is remove `CommonListCollectionsIterable`/`CommonListCollectionsPublisher` from this PR, and leave a comment in the ticket with a link to this discussion.